posts/2021-05-07-caught-red-handed.md

```diff
@@ -0,0 +1,125 @@
+
+[issue-84970]: https://github.com/rust-lang/rust/issues/84970
+
+It turns out `verify-ich` was turned on in version 1.52.0, which was [released recently][].
+
+[released recently]: /2021/05/06/Rust-1.52.0.html
+
+## How does this show up
+
+
+Essentially, for some crates, certain sequences of edit-compile cycles will cause `rustc` to hit the "unstable fingerprints" ICE. I showed one example at the start of this blog post.
+
+Another recent example looks [like this](https://github.com/rust-lang/rust/issues/85039):
+
+```
+thread 'rustc' panicked at 'found unstable fingerprints for predicates_of(<massive text describing rustc internals elided>)', /rustc/88f19c6dab716c6281af7602e30f413e809c5974/compiler/rustc_query_system/src/query/plumbing.rs:593:5
+```
+
+They all arise from inconsistencies when comparing the incremental-compilation cache stored on disk against the values computed during a current `rustc` invocation, which means they all arise from using incremental compilation.
+
+There are three ways that you may have incremental compilation turned on: You may have set the [environment variable][env-vars] `CARGO_INCREMENTAL=1`, or you may have enabled the `build.incremental` [setting in your Cargo.toml][cargo-toml], or you may be building with the `dev` or `test` [profiles][], which default to having incremental compilation enabled.
+
+[env-vars]: https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-reads
+[cargo-toml]: https://doc.rust-lang.org/cargo/reference/config.html#buildincremental
+[profiles]: https://doc.rust-lang.org/cargo/reference/profiles.html
+
+If your project has not opted into enabling incremental compilation, then none of the fingerprint issues should affect your release builds.
+
+## What should a Rust programmer do in response
+
+The Internal Compiler Error asks you to report a bug, and if you can do so, we still want that information. We *want* to know about the cases that are failing.
+
+But regardless of whether or not you file a bug, the problem here can be resolved by either:
+
+ 1. deleting your incremental compilation cache (e.g. by running `cargo clean`), or
+ 2. force incremental compilation to be disabled, by setting `CARGO_INCREMENTAL=0` in your environment or `build.incremental` to `false` in the `config.toml`.
+
+We recommend that users of 1.52.0 disable incremental compilation, to avoid running into this problem.
+
+We do *not* recommend that users of 1.52.0 downgrade to an earlier version of Rust in response to this problem. As noted above, there is at least one instance of a silent [miscompilation][issue-82920] caused by incremental compilation that was not caught until we added the fingerprint checking.
+
+## What is the Rust project going to do to fix this
```

Comment:
I'm not sure if yinz have had time to do this analysis yet, but I'd love to see some reflection in this section on "how did this get missed in nightly/beta and how would we prevent similar mistakes in the future", or an indication that this analysis will be done and published after the fixes are out.

Reply:
I'll provide some background and maybe you can help indicate which of this is useful/helpful information - it may be published in a retrospective later, but I don't know yet if that'll happen.

On the day of the release I submitted https://github.com/rust-lang/rust/issues/84970 to resolve the question of what to do on 1.53 and 1.54 releases. The verification change landed in 1.52, but I had mistakenly assumed that it had landed in 1.53 and we had an additional 6 weeks to either resolve it or improve our messaging; so, ultimately, the fact that this ICE existed *was* known for 12 weeks now, and indeed a few people have been driving the fixing. I think it was not as well understood how common it was - to my knowledge, we had not received reports until after the stable release of it occurring on *every* build, for example. So it was a priority, but not too strongly so.

Another part of the reason for the lack of prioritization is that this is a longstanding bug - to my knowledge, perhaps even to the initial addition of incremental (at least the red/green system, added a few years back), I don't think I'm aware of work since then that would have caused this. So, while it is true that the ICEs are definitely annoying, they merely reveal underlying problems - and in some sense, are better than nothing. I think everyone is in agreement that they are not the best user experience, and that we should avoid them if at all possible, but I think no one has come up with a straightforward solution to do so. It's possible that an alternative to deploying them would be to disable incremental as a default-on option, and indeed that is under active consideration for next week's point release to mitigate the problem.

In hindsight, it's possible that when we discovered this bug we should have immediately moved to do so. However, disabling such a milestone feature that *does* contribute significant performance improvements to many builds would be a hard sell, especially in light of some builds not triggering the verification panics - and as such being able to compile smoothly, even with the added verification. The tradeoff here is not clear.

I will also share that the release team discussed this last Friday and is expecting to invest in better tooling for tracking bugs (regressions in particular, but more generally too) as they propagate to stable, so that we can have a better awareness and hopefully avoid mistakes of the "woops, I thought this would be in 1.53" variety. This should help with a number of other challenges we've encountered as well. However, I expect this to be a long-term project, not one we can quickly address.

Reply:
> "how did this get missed in nightly/beta and how would we prevent similar mistakes in the future"

There has been bug filing going on basically ever since the ICEs were turned on.

If you check the [list of issues](https://github.com/rust-lang/rust/issues?q=is%3Aissue+is%3Aopen+unstable+fingerprints) linked by @Mark-Simulacrum 's issue, you'll see the [oldest](https://github.com/rust-lang/rust/issues/83115) [issues](https://github.com/rust-lang/rust/issues/83126) having been filed on March 14 to the nightly compiler, with the [PR](https://github.com/rust-lang/rust/pull/83007) that enabled the checking having been merged in March 13, so right after that PR merged. With increased exposure to users, the number of reports increased.

But the bugs existed previously too, just caused silent corruptions and weird bugs instead, which is arguably worse than an explicit error message. To the contrary, turning on this error is "making sure the errors won't be missed any more".

Reply:
Ok, I think the start of a TL;DR on root causes sounds to me like:

- We discovered a silent bug(s) with incremental compilation that have probably existed since incremental compilation was added
- We made the bug into a loud compiler error rather than a silent bug
- While that change was in nightly and beta, we got some bug reports but didn't realize exactly how often this bug would be hit in normal development with stable (most people/projects probably only run night/beta in CI where incremental compilation is either irrelevant or disabled)
- We thought this change was only going into beta this cycle; longer term, we will be looking into tooling improvements to provide more clarity of what version a change is in and possibly how many bugs are being reported about a change as a change switches trains
- It also sounds like we should do some thinking on how to get more user testing of parts of Rust only used during development while they're on nightly and beta?

Please feel free to edit this if I've misunderstood or if you disagree with my conclusions or save this stuff for a retro, but as a Rust user reading this, I mostly want to know that the team is going to take a bigger view of this as well as fixing the immediate issue.